File: python/mosna_xy/spec.py
"""Reading what Rust wrote.

A specification is a folder: one `figure.json`, and beside it the binary blobs
its larger arrays were written to. Everything is read through `Spec`, so a
malformed document is refused in one place, by name, rather than surfacing
three modules later as an unhelpful `KeyError`.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Sequence

import numpy as np

#: The name of the document inside a queued figure's folder.
DOCUMENT_NAME = "figure.json"

#: How a blob's declared type maps onto a numpy one. Little-endian explicitly:
#: the file is read on the machine that wrote it in every case that matters,
#: but "in every case that matters" is not a guarantee, and a big-endian reader
#: silently producing garbage coordinates is the worst possible failure.
DTYPES = {"f64": "<f8", "u32": "<u4"}


class SpecError(ValueError):
    """A specification that cannot be drawn from."""


@dataclass(frozen=True)
class Spec:
    """One figure to draw."""

    kind: str
    stem: str
    save_dir: Path
    body: dict[str, Any]
    directory: Path

    @classmethod
    def load(cls, path: Path | str) -> "Spec":
        """Read one `figure.json`."""
        path = Path(path)
        try:
            body = json.loads(path.read_text())
        except OSError as error:
            raise SpecError(f"cannot read {path}: {error}") from error
        except json.JSONDecodeError as error:
            raise SpecError(f"{path} is not valid JSON: {error}") from error

        if not isinstance(body, dict):
            raise SpecError(f"{path} is not a figure: expected an object")

        for field in ("kind", "stem", "save_dir"):
            if not isinstance(body.get(field), str) or not body[field]:
                raise SpecError(f"{path} has no {field}")

        stem = body["stem"]
        # A stem is a file name and nothing more. It is written by this
        # application, so this cannot currently fire — which is exactly when a
        # check is cheap, and the failure it prevents is a figure written
        # outside the directory the user is looking at.
        if Path(stem).name != stem or stem in {".", ".."}:
            raise SpecError(f"{path} has a stem that is not a file name: {stem!r}")

        return cls(
            kind=body["kind"],
            stem=stem,
            save_dir=Path(body["save_dir"]),
            body=body,
            directory=path.parent,
        )

    def array(self, key: str, dtype: str = "f64") -> np.ndarray:
        """The array at `key`: a blob beside the document, or a list inside it.

        An absent key is an empty array. A figure whose data did not survive
        the analysis draws nothing; it does not take down the run that produced
        it, an hour of computation in.
        """
        value = self.body.get(key)
        if value is None:
            return np.empty(0, dtype=DTYPES.get(dtype, "<f8"))

        if isinstance(value, dict) and "__blob__" in value:
            return self._blob(key, value)

        try:
            return np.asarray(value, dtype=DTYPES.get(dtype, "<f8"))
        except (TypeError, ValueError) as error:
            raise SpecError(f"{key} is not an array: {error}") from error

    def _blob(self, key: str, reference: dict[str, Any]) -> np.ndarray:
        name = reference.get("__blob__")
        declared = reference.get("dtype", "f64")
        shape = tuple(int(n) for n in reference.get("shape", []))

        if declared not in DTYPES:
            raise SpecError(f"{key} has an unknown type {declared!r}")

        path = self.directory / str(name)
        try:
            values = np.fromfile(path, dtype=DTYPES[declared])
        except OSError as error:
            raise SpecError(f"cannot read the blob {name} of {key}: {error}") from error

        expected = int(np.prod(shape)) if shape else values.size
        if values.size != expected:
            # Reshaping what fits would draw a plausible picture of the wrong
            # data, which is worse than not drawing one.
            raise SpecError(
                f"the blob {name} of {key} holds {values.size} values, "
                f"but its shape {shape} needs {expected}"
            )
        return values.reshape(shape) if shape else values

    def strings(self, key: str, default: Sequence[str] | None = None) -> list[str]:
        """The list of names at `key`."""
        value = self.body.get(key)
        if value is None:
            return list(default or [])
        if not isinstance(value, list):
            raise SpecError(f"{key} is not a list of names")
        return [str(item) for item in value]

    def text(self, key: str, default: str = "") -> str:
        """The string at `key`."""
        value = self.body.get(key)
        return default if value is None else str(value)

    def number(self, key: str, default: float = 0.0) -> float:
        """The number at `key`."""
        value = self.body.get(key)
        if value is None:
            return default
        try:
            return float(value)
        except (TypeError, ValueError) as error:
            raise SpecError(f"{key} is not a number: {value!r}") from error

    def flag(self, key: str, default: bool = False) -> bool:
        """The boolean at `key`."""
        value = self.body.get(key)
        return default if value is None else bool(value)

    def output(self, extension: str) -> Path:
        """Where the figure is written, in one format."""
        return self.save_dir / f"{self.stem}.{extension}"

File: crates/mosna-xy/src/spec.rs
//! What Rust hands the Python renderer.
//!
//! A specification is *everything* the figure needs and nothing the renderer
//! has to work out for itself: the values, the labels, the colours, the title,
//! the file it is written to. The scientific decisions — which colour map, how
//! a z-score is normalised, in what order the leaves of a dendrogram fall —
//! stay in Rust, where they are already pinned by tests. Python composes an
//! `xy` chart out of what it is given and exports it.
//!
//! # Why a file and not a pipe
//!
//! A queued specification is an artefact that survives the process that wrote
//! it: a figure that came out wrong can be re-rendered from the exact input
//! that produced it, without re-running the analysis that took an hour. The
//! queue is deleted once the renderer has succeeded.

use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicUsize, Ordering};

use serde_json::{Map, Value};

/// A large array, written beside its specification rather than inside it.
///
/// A hundred thousand cells' coordinates are two hundred thousand doubles.
/// Spelled out as JSON text that is some four megabytes to write and to parse,
/// against 1.6 to read with `numpy.fromfile`. Small arrays stay inline, where
/// they are readable.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Blob {
    pub file_name: String,
    pub bytes: Vec<u8>,
}

/// One figure, ready to be drawn.
#[derive(Debug, Clone)]
pub struct Spec {
    kind: String,
    stem: String,
    save_dir: PathBuf,
    body: Map<String, Value>,
    blobs: Vec<Blob>,
}

impl Spec {
    /// A specification of `kind`, to be written as `{stem}.png` and
    /// `{stem}.html` inside `save_dir`.
    pub fn new(kind: &str, stem: impl Into<String>, save_dir: &Path) -> Self {
        Self {
            kind: kind.to_string(),
            stem: stem.into(),
            save_dir: save_dir.to_path_buf(),
            body: Map::new(),
            blobs: Vec::new(),
        }
    }

    /// The kind of figure, which is what the renderer dispatches on.
    pub fn kind(&self) -> &str {
        &self.kind
    }

    /// The file stem the outputs are named after.
    pub fn stem(&self) -> &str {
        &self.stem
    }

    /// Add a field to the body.
    pub fn set(mut self, key: &str, value: impl Into<Value>) -> Self {
        self.body.insert(key.to_string(), value.into());
        self
    }

    /// Add an array of doubles as a binary blob, referenced by name.
    ///
    /// `shape` is the array's shape as `numpy` will read it, so a list of
    /// points is `[n, 2]` and a flat series is `[n]`.
    pub fn set_f64_blob(self, key: &str, values: &[f64], shape: &[usize]) -> Self {
        let bytes = values.iter().flat_map(|v| v.to_le_bytes()).collect();
        self.set_blob(key, bytes, "f64", shape)
    }

    /// Add an array of unsigned 32-bit integers as a binary blob.
    pub fn set_u32_blob(self, key: &str, values: &[u32], shape: &[usize]) -> Self {
        let bytes = values.iter().flat_map(|v| v.to_le_bytes()).collect();
        self.set_blob(key, bytes, "u32", shape)
    }

    fn set_blob(mut self, key: &str, bytes: Vec<u8>, dtype: &str, shape: &[usize]) -> Self {
        let file_name = format!("{key}.bin");
        self.body.insert(
            key.to_string(),
            serde_json::json!({
                "__blob__": file_name,
                "dtype": dtype,
                "shape": shape,
            }),
        );
        self.blobs.push(Blob { file_name, bytes });
        self
    }

    /// The values of a double blob, for the tests that pin what was written.
    ///
    /// The blobs are the part of a specification a reader cannot check by
    /// eye, which is exactly the part worth asserting on.
    pub fn blob_values(&self, key: &str) -> Vec<f64> {
        let name = format!("{key}.bin");
        self.blobs
            .iter()
            .find(|blob| blob.file_name == name)
            .map(|blob| {
                blob.bytes
                    .chunks_exact(8)
                    .map(|chunk| f64::from_le_bytes(chunk.try_into().unwrap()))
                    .collect()
            })
            .unwrap_or_default()
    }

    /// The document the renderer parses.
    ///
    /// The blob file names inside it are relative to the specification's own
    /// directory, so a queue can be moved wholesale without rewriting it.
    pub fn to_json(&self) -> Value {
        let mut document = self.body.clone();
        document.insert("kind".into(), Value::String(self.kind.clone()));
        document.insert("stem".into(), Value::String(self.stem.clone()));
        document.insert(
            "save_dir".into(),
            Value::String(self.save_dir.display().to_string()),
        );
        Value::Object(document)
    }
}

/// Where queued specifications live under a working directory.
///
/// A dot-prefixed name, and one the interface's image scan never descends
/// into: a queue that showed up in the gallery as a folder of nothing would be
/// a bug report.
pub const QUEUE_DIRECTORY: &str = ".mosna-figures";

/// The specifications waiting to be drawn.
///
/// Thread-safe: the analyses draw from inside a `rayon` loop, and two samples
/// finishing at once must not be given the same sequence number — that would
/// silently drop one of the two figures.
#[derive(Debug)]
pub struct Queue {
    directory: PathBuf,
    next: AtomicUsize,
}

impl Queue {
    /// A queue under `working_dir`, created on first use.
    pub fn new(working_dir: &Path) -> Self {
        Self {
            directory: working_dir.join(QUEUE_DIRECTORY),
            next: AtomicUsize::new(0),
        }
    }

    /// Where the specifications are written.
    pub fn directory(&self) -> &Path {
        &self.directory
    }

    /// How many specifications have been queued.
    pub fn len(&self) -> usize {
        self.next.load(Ordering::SeqCst)
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Write one specification, and its blobs, into the queue.
    ///
    /// Returns the path of the document written.
    pub fn push(&self, spec: Spec) -> anyhow::Result<PathBuf> {
        let sequence = self.next.fetch_add(1, Ordering::SeqCst);
        let folder = self.directory.join(format!("{sequence:05}-{}", spec.kind));
        std::fs::create_dir_all(&folder)
            .map_err(|e| anyhow::anyhow!("cannot create {}: {e}", folder.display()))?;

        for blob in &spec.blobs {
            let path = folder.join(&blob.file_name);
            std::fs::write(&path, &blob.bytes)
                .map_err(|e| anyhow::anyhow!("cannot write {}: {e}", path.display()))?;
        }

        let path = folder.join("figure.json");
        let document = serde_json::to_vec_pretty(&spec.to_json())?;
        std::fs::write(&path, document)
            .map_err(|e| anyhow::anyhow!("cannot write {}: {e}", path.display()))?;
        Ok(path)
    }

    /// Remove the queue once its figures have been drawn.
    ///
    /// Failure to clean up is not a failure of the analysis: the figures are
    /// on disk, and a leftover queue costs disk space and nothing else. It is
    /// reported by the caller, not raised.
    pub fn discard(&self) -> std::io::Result<()> {
        match std::fs::remove_dir_all(&self.directory) {
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(()),
            other => other,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_specification_carries_its_kind_its_stem_and_its_directory() {
        let spec = Spec::new("embedding", "cluster_labels", Path::new("/runs/niches"));
        let json = spec.to_json();
        assert_eq!(json["kind"], "embedding");
        assert_eq!(json["stem"], "cluster_labels");
        assert_eq!(json["save_dir"], "/runs/niches");
    }

    #[test]
    fn fields_reach_the_document_unchanged() {
        let spec = Spec::new("histogram", "h", Path::new("/tmp"))
            .set("title", "Niches")
            .set("counts", serde_json::json!([1, 2, 3]));
        let json = spec.to_json();
        assert_eq!(json["title"], "Niches");
        assert_eq!(json["counts"], serde_json::json!([1, 2, 3]));
    }

    /// The bytes have to be little-endian doubles in row-major order, because
    /// that is what `numpy.fromfile` reads back without being told anything.
    #[test]
    fn a_double_blob_is_little_endian_and_row_major() {
        let spec = Spec::new("network", "net_1", Path::new("/tmp")).set_f64_blob(
            "coords",
            &[1.0, 2.0, 3.0, 4.0],
            &[2, 2],
        );

        assert_eq!(spec.to_json()["coords"]["dtype"], "f64");
        assert_eq!(spec.to_json()["coords"]["shape"], serde_json::json!([2, 2]));
        assert_eq!(spec.to_json()["coords"]["__blob__"], "coords.bin");

        let blob = &spec.blobs[0];
        assert_eq!(blob.bytes.len(), 4 * 8);
        assert_eq!(&blob.bytes[..8], &1.0f64.to_le_bytes());
        assert_eq!(&blob.bytes[8..16], &2.0f64.to_le_bytes());
    }

    #[test]
    fn an_integer_blob_is_four_bytes_a_value() {
        let spec = Spec::new("network", "net_1", Path::new("/tmp")).set_u32_blob(
            "edges",
            &[0, 1],
            &[1, 2],
        );
        assert_eq!(spec.blobs[0].bytes.len(), 2 * 4);
        assert_eq!(spec.to_json()["edges"]["dtype"], "u32");
    }

    #[test]
    fn an_empty_array_is_still_a_blob_the_renderer_can_read() {
        let spec =
            Spec::new("network", "net_1", Path::new("/tmp")).set_f64_blob("coords", &[], &[0, 2]);
        assert_eq!(spec.blobs[0].bytes.len(), 0);
        assert_eq!(spec.to_json()["coords"]["shape"], serde_json::json!([0, 2]));
    }

    #[test]
    fn pushing_writes_the_document_and_its_blobs_together() {
        let dir = tempfile::tempdir().unwrap();
        let queue = Queue::new(dir.path());
        let spec =
            Spec::new("network", "net_1", dir.path()).set_f64_blob("coords", &[1.0, 2.0], &[1, 2]);

        let path = queue.push(spec).unwrap();
        assert!(path.is_file(), "{} was not written", path.display());
        assert!(path.parent().unwrap().join("coords.bin").is_file());

        let document: Value = serde_json::from_slice(&std::fs::read(&path).unwrap()).unwrap();
        assert_eq!(document["kind"], "network");
    }

    /// Two samples finishing at the same instant must not overwrite one
    /// another; the sequence number is what keeps them apart, and it is also
    /// what fixes the order the renderer works through.
    #[test]
    fn concurrent_pushes_each_get_their_own_folder() {
        let dir = tempfile::tempdir().unwrap();
        let queue = Queue::new(dir.path());

        let root = dir.path();
        std::thread::scope(|scope| {
            for index in 0..16 {
                let queue = &queue;
                scope.spawn(move || {
                    queue
                        .push(Spec::new("network", format!("net_{index}"), root))
                        .unwrap();
                });
            }
        });

        let folders: Vec<_> = std::fs::read_dir(queue.directory())
            .unwrap()
            .flatten()
            .collect();
        assert_eq!(folders.len(), 16);
        assert_eq!(queue.len(), 16);
    }

    #[test]
    fn the_queue_is_named_so_the_gallery_never_shows_it() {
        let queue = Queue::new(Path::new("/runs"));
        assert_eq!(queue.directory(), Path::new("/runs/.mosna-figures"));
        assert!(QUEUE_DIRECTORY.starts_with('.'));
    }

    #[test]
    fn discarding_removes_the_queue_and_tolerates_its_absence() {
        let dir = tempfile::tempdir().unwrap();
        let queue = Queue::new(dir.path());
        queue.push(Spec::new("histogram", "h", dir.path())).unwrap();
        assert!(queue.directory().is_dir());

        queue.discard().unwrap();
        assert!(!queue.directory().exists());
        queue.discard().expect("discarding twice is not an error");
    }
}
